File: java/src/main/java/dev/tensorhero/tinytorch/Trainer.java
package dev.tensorhero.tinytorch;

import dev.tensorhero.tinynum.NDArray;

import java.util.List;
import java.util.function.BiFunction;

/**
 * Trainer — orchestrates the standard training loop.
 *
 * <p>Encapsulates forward → loss → backward → step → zeroGrad,
 * plus utilities for learning rate scheduling, gradient clipping,
 * and accuracy evaluation.</p>
 */
public class Trainer {

    private Layer model;
    private Optimizer optimizer;
    private BiFunction<Tensor, Tensor, Tensor> lossFunction;

    // ================================================================
    // E09 — Training Loop
    // ================================================================

    public Trainer(Layer model, Optimizer optimizer,
                   BiFunction<Tensor, Tensor, Tensor> lossFunction) {
        throw new UnsupportedOperationException("TODO: E09");
    }

    /**
     * Executes one training step: forward → loss → backward → step → zeroGrad.
     *
     * @return the scalar loss value
     */
    public float trainStep(Tensor x, Tensor y) {
        throw new UnsupportedOperationException("TODO: E09");
    }

    /**
     * Computes classification accuracy by comparing argmax of pred and target rows.
     *
     * @param pred   model output, shape [batch, numClasses]
     * @param target one-hot encoded labels, shape [batch, numClasses]
     * @return accuracy in [0, 1]
     */
    public static float accuracy(Tensor pred, Tensor target) {
        throw new UnsupportedOperationException("TODO: E09");
    }

    /**
     * Clips gradient norm. If global L2 norm exceeds maxNorm, scales all gradients down.
     *
     * @param params  list of parameters whose gradients to clip
     * @param maxNorm maximum allowed gradient norm
     * @return the original (pre-clip) gradient norm
     */
    public static float clipGradNorm(List<Tensor> params, float maxNorm) {
        throw new UnsupportedOperationException("TODO: E09");
    }

    /**
     * Cosine annealing learning rate schedule.
     *
     * @return learning rate at the given step
     */
    public static float cosineSchedule(int step, int totalSteps, float maxLr, float minLr) {
        throw new UnsupportedOperationException("TODO: E09");
    }
}

File: python/tinytorch/trainer.py
"""Trainer — orchestrates the standard training loop.

Encapsulates forward → loss → backward → step → zero_grad,
plus utilities for learning rate scheduling, gradient clipping,
and accuracy evaluation.
"""

from __future__ import annotations

import math
from typing import Callable

from tinynum import NDArray
from tinytorch.tensor import Tensor


class Trainer:
    """Training loop orchestrator."""

    # ================================================================
    # E09 — Training Loop
    # ================================================================

    def __init__(
        self,
        model,
        optimizer,
        loss_fn: Callable[[Tensor, Tensor], Tensor],
    ) -> None:
        raise NotImplementedError("TODO: E09")

    def train_step(self, x: Tensor, y: Tensor) -> float:
        """Single batch training step: forward → loss → backward → step → zero_grad."""
        raise NotImplementedError("TODO: E09")

    @staticmethod
    def accuracy(pred: Tensor, target: Tensor) -> float:
        """Computes classification accuracy (argmax comparison)."""
        raise NotImplementedError("TODO: E09")

    @staticmethod
    def clip_grad_norm(params: list[Tensor], max_norm: float) -> float:
        """Clips gradient norm, returns original total norm."""
        raise NotImplementedError("TODO: E09")

    @staticmethod
    def cosine_schedule(step: int, total_steps: int, max_lr: float, min_lr: float) -> float:
        """Cosine annealing learning rate schedule."""
        raise NotImplementedError("TODO: E09")
